language/move-vm/runtime/src/loader.rs

```diff
@@ -169,18 +183,29 @@ impl ModuleCache {
         &mut self,
         natives: &NativeFunctions,
         id: ModuleId,
-        module: CompiledModule,
+        module: &CompiledModule,
     ) -> VMResult<Arc<Module>> {
-        if let Some(cached) = self.module_at(&id) {
+        if let Some(cached) = self.loaded_module_at(&id) {
             return Ok(cached);
         }
 
+        // Make sure the modules of dependencies are in the cache.
+        for dep in module.immediate_dependencies() {
+            let compiled_dep = self.compiled_module_at(&dep).ok_or_else(|| {
+                PartialVMError::new(StatusCode::MISSING_DEPENDENCY)
+                    .with_message(format!("Cannot load dependency {dep} of {id}"))
+                    .finish(Location::Undefined)
+            })?;
+
+            self.insert(natives, dep, compiled_dep.as_ref())?;
+        }
+
```

Comment:
is all of this under lock now?

Reply:
Yes, it is.  The same work was happening under the write lock before, but we would release and re-acquire the lock between each `insert` of a dependency module.